compiler/IREmitter/Payload/patches/vm_insnhelper.c

```diff
@@ -481,3 +484,232 @@ void sorbet_throwReturn(rb_execution_context_t *ec, VALUE retval) {
+    int offset = computeLocalIndex(index);
+    return *(vm_get_ep(cfp->ep, level) - offset);
+}
+SORBET_INLINE
+static void sorbet_writeLocal(rb_control_frame_t *cfp, long index, long level, VALUE value) {
+    int offset = computeLocalIndex(index);
+    vm_env_write(vm_get_ep(cfp->ep, level), -offset, value);
+}
+
+typedef VALUE (*ExceptionFFIType)(VALUE **pc, VALUE closure, rb_control_frame_t *);
+
+SORBET_INLINE
+static void sorbet_raiseIfNotNil(VALUE exception) {
+    if (exception == RUBY_Qnil) {
+        return;
+    }
+
+    rb_exc_raise(exception);
+}
+VALUE sorbet_run_exception_handling(rb_execution_context_t * volatile ec,
+                                    volatile ExceptionFFIType body,
+                                    VALUE ** volatile pc,
+                                    // The locals offset for the body.
+                                    volatile VALUE methodClosure,
+                                    rb_control_frame_t * volatile cfp,
+                                    // May be nullptr.
+                                    volatile ExceptionFFIType handlers,
+                                    // May be nullptr.
+                                    volatile ExceptionFFIType elseClause,
+                                    // May be nullptr.
+                                    volatile ExceptionFFIType ensureClause,
+                                    // The special value indicating that we need to retry.
+                                    volatile VALUE retrySingleton,
+                                    volatile long exceptionValueIndex,
+                                    volatile long exceptionValueLevel) {
+    // `volatile` is not used in polite C programming, but here it's very important:
+    // it ensures that the requisite variables are stored in memory across the setjmp
+    // below and therefore will still be valid upon a return via longjmp.  All such
+    // variables that need to be live "on both sides" of the setjmp need to be
+    // declared with `volatile`.
+    //
+    // Temporary variables used on the "longjmp side" only can be declared as normal.
+    volatile VALUE executionResult = Qundef;
+    volatile VALUE previousException = ec->errinfo;
+    volatile VALUE bodyException = Qnil;
+    volatile VALUE handlerException = Qnil;
+    volatile enum {
+        RunningBody,
+        RunningHandlers,
+    } state = RunningBody;
+    volatile enum ruby_tag_type nleType;
+
+    EC_PUSH_TAG(ec);
+
+    if ((nleType = EC_EXEC_TAG()) == TAG_NONE) {
+    execute_body:
+        state = RunningBody;
+        // tag.state will have been reset appropriately if we got here via `retry`,
+        // but not this variable.
+        nleType = TAG_NONE;
+
+        // Clear out the local variable shared across exception handling regions
+        // where we store the current exception value.
+        sorbet_writeLocal(cfp, exceptionValueIndex, exceptionValueLevel, Qnil);
```

Comment:
Because it's the sort of optimization I can imagine us wanting us to write in the future: could you write a comment here saying that it's necessary to use `sorbet_writeLocal` here, as the locals could move to the heap between runs of this statement? I can imagine us wanting to refactor this to write to a pointer instead, and that's not going to be correct.